src/sentry/rules/processing/processor.py

```diff
@@ -277,30 +309,61 @@ def apply_rule(self, rule: Rule, status: GroupRuleStatus) -> None:
-
-        # Sort `condition_list` so that most expensive conditions run last.
-        condition_list.sort(key=lambda condition: is_condition_slow(condition))
-
-        for predicate_list, match, name in (
-            (filter_list, filter_match, "filter"),
-            (condition_list, condition_match, "condition"),
-        ):
-            if not predicate_list:
-                continue
-            predicate_iter = (self.condition_matches(f, state, rule) for f in predicate_list)
-            predicate_func = get_match_function(match)
+        fast_conditions, slow_conditions = self.group_conditions_by_speed(condition_list)
+        process_slow_conditions_later = features.has(
+            "organizations:process-slow-alerts", self.project.organization
+        )
+        condition_list = fast_conditions
+        if not process_slow_conditions_later:
+            condition_list = fast_conditions + slow_conditions  # type: ignore[operator]
+
+        # evaluate all filters and return if they fail, then do the enqueue logic for conditions
+        if filter_list:
+            predicate_iter = (self.condition_matches(f, state, rule) for f in filter_list)
+            predicate_func = get_match_function(filter_match)
             if predicate_func:
                 if not predicate_func(predicate_iter):
                     return
             else:
-                log_string = f"Unsupported {name}_match {match!r} for rule {rule.id}"
+                log_string = f"Unsupported filter_match {filter_match} for rule {rule.id}"
+                logger.error(
+                    log_string,
+                    filter_match,
+                    rule.id,
+                    extra={**logging_details},
+                )
+                return
+
+        predicate_func = get_match_function(condition_match)
+        if condition_list:
+            if not predicate_func:
+                log_string = f"Unsupported condition_match {condition_match} for rule {rule.id}"
                 logger.error(
                     log_string,
                     filter_match,
                     rule.id,
                     extra={**logging_details},
                 )
                 return
+            predicate_iter = (self.condition_matches(f, state, rule) for f in condition_list)
+            result = predicate_func(predicate_iter)
+
+            if condition_match == "any" and not result:
+                if slow_conditions and process_slow_conditions_later:
+                    self.enqueue_rule(rule)
+                return
+
+            elif condition_match == "all":
+                if not result:
+                    return
+
+                if slow_conditions and process_slow_conditions_later:
+                    self.enqueue_rule(rule)
+                    return
+
+        if slow_conditions and process_slow_conditions_later:
+            self.enqueue_rule(rule)
+            return
```

Comment:
I feel like there are cases where this will get hit where we don't want it to... Should it be in the `else` for `if condition_list`?

Like what if we have 
```
match = "any"
fast_conditions = [c1-pass]
slow_conditions = [c2]
```

So we'll go through the if statement, then `c1-pass` gets a True result so we don't return. But then we'll hit this condition and enqueue, even though the condition has already passed

Reply:
Hmm it can't be the else because if we have the flag, slow and fast conditions with an any match, and the fast condition fails, we won't enqueue it. Maybe I need some new boolean that defaults to `False` and gets set to `True` if it's an any match, we have the flag and a fast condition, and then if True we enqueue if there are slow conditions?